File: swift/Shared/JefRenderer.swift
import Foundation
import CoreGraphics
import ImageIO
import os

let jefLog = Logger(subsystem: "com.flowdev.jefpreviwe", category: "render")

enum JefRenderError: LocalizedError {
    case parseFailed(String)
    case renderFailed(String)

    var errorDescription: String? {
        switch self {
        case .parseFailed(let m): return "JEF parse failed: \(m)"
        case .renderFailed(let m): return "JEF render failed: \(m)"
        }
    }
}

enum JefBackground { case transparent, white }

/// Нативный рендер .jef → CGImage. Порт jef_render.py (без Python-подпроцесса —
/// QuickLook-расширения не могут порождать процессы в своей sandbox).
struct JefRenderer {

    static let supersample = 2
    // Запасная палитра, если в файле нет данных о нитях (как в jef_render.py).
    static let fallback: [UInt32] = [
        0x1A76BC, 0xD83A3E, 0x34A048, 0xF0921F, 0x7E57C2, 0x009688,
    ]

    static func renderImage(fileURL: URL, size: Int, background: JefBackground,
                            lineWidthMM: CGFloat = 2.2, drawJumps: Bool = false) throws -> CGImage {
        let data = try Data(contentsOf: fileURL)
        let pattern = try JefParser.parse(data)
        jefLog.log("parsed: stitches=\(pattern.stitches.count, privacy: .public) colors=\(pattern.colors.count, privacy: .public)")
        return try draw(pattern, size: size, background: background,
                        lineWidthMM: lineWidthMM, drawJumps: drawJumps)
    }

    static func renderPNG(fileURL: URL, size: Int, background: JefBackground) throws -> Data {
        let img = try renderImage(fileURL: fileURL, size: size, background: background)
        let out = NSMutableData()
        guard let dest = CGImageDestinationCreateWithData(
            out as CFMutableData, "public.png" as CFString, 1, nil
        ) else { throw JefRenderError.renderFailed("не удалось создать PNG-назначение") }
        CGImageDestinationAddImage(dest, img, nil)
        guard CGImageDestinationFinalize(dest) else {
            throw JefRenderError.renderFailed("не удалось сериализовать PNG")
        }
        return out as Data
    }

    @inline(__always) private static func rgb(_ hex: UInt32) -> (CGFloat, CGFloat, CGFloat) {
        (CGFloat((hex >> 16) & 0xFF) / 255.0,
         CGFloat((hex >> 8) & 0xFF) / 255.0,
         CGFloat(hex & 0xFF) / 255.0)
    }

    private static func draw(_ pat: JefPattern, size: Int, background: JefBackground,
                             lineWidthMM: CGFloat, drawJumps: Bool) throws -> CGImage {
        let canvas = size * supersample
        let cd = Double(canvas)
        let w = max(pat.maxX - pat.minX, 1)
        let h = max(pat.maxY - pat.minY, 1)
        let pad = 0.05 * cd
        let scale = min((cd - 2 * pad) / w, (cd - 2 * pad) / h)
        let ox = (cd - w * scale) / 2 - pat.minX * scale
        let oy = (cd - h * scale) / 2 - pat.minY * scale
        @inline(__always) func px(_ x: Double, _ y: Double) -> CGPoint {
            CGPoint(x: x * scale + ox, y: y * scale + oy)
        }

        let cs = CGColorSpaceCreateDeviceRGB()
        let bmp = CGImageAlphaInfo.premultipliedLast.rawValue
        guard let ctx = CGContext(data: nil, width: canvas, height: canvas,
                                  bitsPerComponent: 8, bytesPerRow: 0,
                                  space: cs, bitmapInfo: bmp) else {
            throw JefRenderError.renderFailed("не удалось создать графический контекст")
        }

        if case .white = background {
            ctx.setFillColor(red: 1, green: 1, blue: 1, alpha: 1)
            ctx.fill(CGRect(x: 0, y: 0, width: canvas, height: canvas))
        }

        // Ориентация как в PIL (origin сверху-слева): переворачиваем ось Y.
        ctx.translateBy(x: 0, y: cd)
        ctx.scaleBy(x: 1, y: -1)
        ctx.setLineJoin(.round)
        ctx.setLineCap(.round)
        let lw = max(1.0, (lineWidthMM * CGFloat(supersample) * CGFloat(size) / 1024.0).rounded())
        ctx.setLineWidth(lw)

        let colors = pat.colors
        func colorFor(_ i: Int) -> (CGFloat, CGFloat, CGFloat) {
            if !colors.isEmpty { return rgb(colors[i % colors.count]) }
            return rgb(fallback[i % fallback.count])
        }

        var idx = 0
        var seg: [CGPoint] = []
        func flush() {
            if seg.count >= 2 {
                let c = colorFor(idx)
                ctx.setStrokeColor(red: c.0, green: c.1, blue: c.2, alpha: 1)
                ctx.beginPath()
                ctx.addLines(between: seg)
                ctx.strokePath()
            }
            seg.removeAll(keepingCapacity: true)
        }

        for s in pat.stitches {
            switch s.cmd {
            case .stitch:
                seg.append(px(s.x, s.y))
            case .jump:
                if drawJumps && !seg.isEmpty { seg.append(px(s.x, s.y)) }
                flush()
                seg = [px(s.x, s.y)]
            case .colorChange:
                flush(); idx += 1
            case .stop:
                flush()
            case .end:
                flush()
            }
        }
        flush()

        guard let full = ctx.makeImage() else {
            throw JefRenderError.renderFailed("makeImage вернул nil")
        }
        // Даунсэмплинг supersample→size с качественной интерполяцией (антиалиасинг).
        guard let down = CGContext(data: nil, width: size, height: size,
                                   bitsPerComponent: 8, bytesPerRow: 0,
                                   space: cs, bitmapInfo: bmp) else {
            return full
        }
        down.interpolationQuality = .high
        down.draw(full, in: CGRect(x: 0, y: 0, width: size, height: size))
        return down.makeImage() ?? full
    }
}

File: windows/src/JefPreview.Core/JefRenderer.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Drawing.Imaging;
using System.IO;

namespace JefPreview.Core;

public enum JefBackground
{
    Transparent,
    White
}

/// <summary>Нативный рендер .jef → Bitmap. Порт Swift JefRenderer.</summary>
public static class JefRenderer
{
    public const int Supersample = 2;

    private static readonly uint[] Fallback =
    {
        0x1A76BC, 0xD83A3E, 0x34A048, 0xF0921F, 0x7E57C2, 0x009688,
    };

    public static Bitmap RenderImage(string filePath, int size, JefBackground background,
        double lineWidthMm = 2.2, bool drawJumps = false)
    {
        var pattern = JefParser.ParseFile(filePath);
        return Draw(pattern, size, background, lineWidthMm, drawJumps);
    }

    public static Bitmap RenderImage(byte[] data, int size, JefBackground background,
        double lineWidthMm = 2.2, bool drawJumps = false)
    {
        var pattern = JefParser.Parse(data);
        return Draw(pattern, size, background, lineWidthMm, drawJumps);
    }

    public static byte[] RenderPng(string filePath, int size, JefBackground background)
    {
        using var bmp = RenderImage(filePath, size, background);
        using var ms = new MemoryStream();
        bmp.Save(ms, ImageFormat.Png);
        return ms.ToArray();
    }

    private static (byte R, byte G, byte B) Rgb(uint hex) =>
        ((byte)((hex >> 16) & 0xFF), (byte)((hex >> 8) & 0xFF), (byte)(hex & 0xFF));

    private static Bitmap Draw(JefPattern pat, int size, JefBackground background,
        double lineWidthMm, bool drawJumps)
    {
        var canvas = size * Supersample;
        var cd = (double)canvas;
        var w = Math.Max(pat.MaxX - pat.MinX, 1);
        var h = Math.Max(pat.MaxY - pat.MinY, 1);
        var pad = 0.05 * cd;
        var scale = Math.Min((cd - 2 * pad) / w, (cd - 2 * pad) / h);
        var ox = (cd - w * scale) / 2 - pat.MinX * scale;
        var oy = (cd - h * scale) / 2 - pat.MinY * scale;

        PointF Px(double x, double y) =>
            new((float)(x * scale + ox), (float)(y * scale + oy));

        var pixelFormat = background == JefBackground.Transparent
            ? PixelFormat.Format32bppArgb
            : PixelFormat.Format24bppRgb;

        using var full = new Bitmap(canvas, canvas, pixelFormat);
        using (var g = Graphics.FromImage(full))
        {
            g.SmoothingMode = SmoothingMode.HighQuality;
            g.InterpolationMode = InterpolationMode.HighQualityBicubic;
            g.PixelOffsetMode = PixelOffsetMode.HighQuality;

            if (background == JefBackground.White)
                g.Clear(Color.White);
            else
                g.Clear(Color.Transparent);

            var lw = Math.Max(1.0, Math.Round(lineWidthMm * Supersample * size / 1024.0));
            using var pen = new Pen(Color.Black, (float)lw)
            {
                StartCap = LineCap.Round,
                EndCap = LineCap.Round,
                LineJoin = LineJoin.Round
            };

            (byte R, byte G, byte B) ColorFor(int i)
            {
                if (pat.Colors.Count > 0)
                {
                    var c = pat.Colors[i % pat.Colors.Count];
                    return Rgb(c);
                }

                return Rgb(Fallback[i % Fallback.Length]);
            }

            var idx = 0;
            var seg = new List<PointF>();

            void Flush()
            {
                if (seg.Count < 2)
                {
                    seg.Clear();
                    return;
                }

                var c = ColorFor(idx);
                pen.Color = Color.FromArgb(c.R, c.G, c.B);
                g.DrawLines(pen, seg.ToArray());
                seg.Clear();
            }

            foreach (var s in pat.Stitches)
            {
                switch (s.Command)
                {
                    case JefStitchCommand.Stitch:
                        seg.Add(Px(s.X, s.Y));
                        break;
                    case JefStitchCommand.Jump:
                        if (drawJumps && seg.Count > 0)
                            seg.Add(Px(s.X, s.Y));
                        Flush();
                        seg.Add(Px(s.X, s.Y));
                        break;
                    case JefStitchCommand.ColorChange:
                        Flush();
                        idx++;
                        break;
                    case JefStitchCommand.Stop:
                        Flush();
                        break;
                    case JefStitchCommand.End:
                        Flush();
                        break;
                }
            }

            Flush();
        }
        var down = new Bitmap(size, size, pixelFormat);
        using (var g = Graphics.FromImage(down))
        {
            g.InterpolationMode = InterpolationMode.HighQualityBicubic;
            g.SmoothingMode = SmoothingMode.HighQuality;
            g.DrawImage(full, new Rectangle(0, 0, size, size));
        }

        return down;
    }
}
